source/extensions/filters/http/proto_api_scrubber/scrubbing_util_lib/field_checker.cc

```diff
@@ -16,6 +17,121 @@ namespace Extensions {
 namespace HttpFilters {
 namespace ProtoApiScrubber {
 
+namespace {
+
+// Internal state used to track position and results during path traversal.
+// This struct is local to the source file and does not rely on private class members.
+struct TraversalState {
+  const Protobuf::Descriptor* current_desc;
+  std::vector<std::string> normalized_path;
+  bool is_map_entry;
+
+  TraversalState(const Protobuf::Descriptor* root, size_t capacity)
+      : current_desc(root), is_map_entry(false) {
+    normalized_path.reserve(capacity);
+  }
+};
+
+// Updates the descriptor pointer based on the current field.
+// Returns true if traversal can continue (message type), false otherwise.
+inline bool resolveNextMessageDescriptor(const Protobuf::FieldDescriptor* field,
+                                         TraversalState& state) {
+  if (!field || field->type() != Protobuf::FieldDescriptor::TYPE_MESSAGE) {
+    state.current_desc = nullptr;
+    return false;
+  }
+  state.current_desc = field->message_type();
+  return true;
+}
+
+// Checks if the current field is a map and performs normalization (key -> "value") if needed.
+// Updates the traversal index if the map key segment is consumed.
+inline void normalizeMapEntryPath(const Protobuf::FieldDescriptor* field,
+                                  const std::vector<std::string>& path, size_t& index,
+                                  TraversalState& state) {
+  // Check if we are at a Map field and have a subsequent segment (the key) to consume.
+  if (field->is_map() && index + 1 < path.size()) {
+    state.normalized_path.push_back("value");
+
+    // Skip the actual key segment in the input path since we normalized it to "value".
+    index++;
+
+    // If we just consumed the last segment (the key), this path points to a map entry.
+    if (index == path.size() - 1) {
+      state.is_map_entry = true;
+    }
+
+    // Advance the descriptor to the Map Value's type (Field number 2 is always 'value').
+    const auto* value_field = state.current_desc->FindFieldByNumber(2);
+    if (value_field && value_field->type() == Protobuf::FieldDescriptor::TYPE_MESSAGE) {
+      state.current_desc = value_field->message_type();
+    } else {
+      // Value is primitive or enum; we cannot traverse deeper.
+      state.current_desc = nullptr;
+    }
+  }
+}
+
+// Processes a single path segment: updates normalized path and advances descriptor.
```

Comment:
Consider adding a comment that describes this from a high-level perspective.
It's unclear what this method is doing with the in-out arguments.